text/3064-bitfield.md

```diff
@@ -0,0 +1,384 @@
+attribute to fields:
+
+```rust
+#[repr(C)]
+struct X {
+    #[bitfield(1)] a: u8,
+    #[bitfield(8)] b: u8,
+    #[bitfield(0)] _: u8,
+    #[bitfield(2)] d: u8,
+}
+```
+
+corresponds to the C struct
+
+```c
+struct X {
+    unsigned char a: 1;
+    unsigned char b: 8;
+    unsigned char  : 0;
+    unsigned char d: 2;
+};
+```
+
+As you can see in the example, unnamed bit-fields in C are written with the
+reserved identifier `_` in Rust. When you have a struct with an unnamed
+bit-field, you cannot access that field. On the other hand, you also do not have
+to define a value for that field when constructing the struct:
+
+```rust
+let x = X { a: 1, b: 1, d: 1 };
+let X { a, b, d } = x;
+```
+
+Just like in C, you cannot take the address of a bit-field:
+
+```rust
+fn f(x: &X) -> &u8 {
+    &x.a // !!! does not compile
+}
+```
+
+The value of the `N` in `bitfield(N)` must be a non-negative integer literal.
+This literal must not be larger than the bit-size of the type of the field.
+
+In debug builds, when you write to a bit-field, Rust performs overflow checks.
+If the value to be stored does not fit in the bit-field, the overflow check
+fails and the operation panics:
+
+```rust
+fn f(x: &mut X) {
+    x.a = 6; // !!! overflow check fails
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The attribute `bitfield(N)` is added. `N` must be an integer literal. `N` is
+called the *width* of the bit-field. This attribute can only be applied to fields
+of a `repr(C)` struct. Such a field is called a *bit-field*. A bit-field must
+have one of the types
+
+- `bool`,
+- `u8`, `u16`, `u32`, `u64`, `u128`, `usize`,
+- `i8`, `i16`, `i32`, `i64`, `i128`, `isize`
+
+modulo type aliases.
+
+The width of a bit-field with type `T` must be in the range `[0,
+bit_width_of(T)]`. For the integer types, `bit_width_of` is defined as `8 *
+sizeof`. For `bool`, `bit_width_of` is 1.
+
+A bit-field can have the name `_`. Such fields are only used to modify the layout
+of the struct. Consider the example
+
+```rust
+#[repr(C)]
+struct X {
+    a: u32,
+    #[bitfield(6)] _: u32,
+    #[bitfield(1)] b: u32,
+    #[bitfield(5)] _: u32,
+    #[bitfield(1)] c: u32,
+}
+```
+
+This struct behaves like
+
+```rust
+#[repr(C)]
+#[insert_unnamed_bitfield_after(0, 6, u32)]
+#[insert_unnamed_bitfield_after(1, 5, u32)]
+struct X {
+    a: u32,
+    #[bitfield(1)] b: u32,
+    #[bitfield(1)] c: u32,
+}
+```
+
+if the attribute `insert_unnamed_bitfield_after` existed and caused the layout
+to be modified in the same way as by the unnamed bit-field as described below.
+
+In particular, such a field cannot be accessed, need not and cannot appear in
+the construction of such a struct, cannot appear in the deconstruction of such a
+struct.
+
+Nevertheless, such fields appear like regular fields in the rustdoc output if
+they have the `pub` visibility.
+
+If the width of the bit-field is `0`, then the name of the field must be `_`.
```

Comment:
Please specify the behavior of `bitfield(0)` in this RFC. In general, I don't have any objection to defining this feature in terms of C-compatible bitfield types. However, I'd like to have a self-contained explanation for `bitfield(0)`.

Reply:
I do not think the behavior of zero-width bit fields is specified enough to make this worth it. I think the most than can be said is that a zero-width bit field causes the next bit field to start at the next 8-bit boundary (the next byte.) But this is probably not what you had in mind.

Reply:
If that is the purpose of a zero width bitfield, maybe it would be better to specify that with an attribute (possibly an argument to the bitfield attribute) on the next bitfield. Zero width bitfield seem very confusing to me.

Reply:
> I do not think the behavior of zero-width bit fields is specified enough to make this worth it.

That, to me, seems like a much *stronger* argument for specifying it explicitly, to make sure that the expected semantic is clear, and to make it feasible to test this.

> I think the most than can be said is that a zero-width bit field causes the storage unit allocated for the next bit field to be aligned at least to the next 8-bit boundary (the next byte).

My understanding of the semantic was, roughly, that a zero-width bitfield will "finish" filling out the current in-progress field of the type specified for the zero-width bitfield.

Reply:
I see no reason to specify this in the RFC. After all, Rust might in the future be ported to systems where the behavior is different. In any case, as per the RFC, the behavior will be described in the reference.

Reply:
I do: I feel that the RFC would be underspecified without it.

The RFC can say that the canonical definition is C-compatible, and then give the behavior for *current* platforms, with a note that if a target's C-compatible bitfield definition differs Rust will do the same. That will still help the definition be clearer and more self-contained.

Putting the definition in the reference seems like a good idea, but I think the RFC should have at least draft text specifying this.

Reply:
> and then give the behavior for _current_ platforms

The behavior for current platforms differs widely. I'm currently working on another RFC that will describe the behavior of repr(C) in detail. Let's defer this point until I've posted that RFC.

Reply:
The Rust Reference prefers that there is a canonical decision made about its contents and not to be the place where those decisions are made.

Reply:
It's not supposed to make a decision only to document it. The decision has already be made by the system compiler of the target. That's msvc on the msvc targets, gcc on the mingw and Linux targets, and clang otherwise.

Reply:
I've written a [crate](https://crates.io/crates/repc) that implements the layout algorithms used my msvc, gcc, and clang. The code contains a large number of test cases that compare its output to the output of the C compilers. It should therefore have a high accuracy.

These layout algorithms are far too complicated to be described in prose. The reference can refer to this crate instead.